pandas/core/internals.py

```diff
@@ -1269,6 +1269,11 @@ def take_nd(self, indexer, axis, new_mgr_locs=None, fill_tuple=None):
 
     def diff(self, n, axis=1, mgr=None):
         """ return block for the diff of the values """
+        if isinstance(self, DatetimeTZBlock) and axis == 0:
+            # This method will iterate (self.apply) over each DateTimeTZBlock
+            # but axis=0 will need to gather data from all DateTimeTZBlocks
```

Comment:
this is not a good long term soln, but you can put a diff method in DatetimeTZBlock to handle this, kind of like we do for shift. Its not great, but this whole diff / shift thing is pretty fundamentally broken ATM. The routines are not general enough to graciously handle the various types of inputs (e.g. 2-D arrays, 1-D arrays, 1-D non-arrays). Further algos.diff hasn't seen much attention to handle tz-aware things. E.g. you have to have a uniform tz in order to diff in the first place.